### PR TITLE
[server] use org service in iam-session-app

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

Replaces use of teamDb with organizationService in iam-session.app.ts

### <samp>🤖 Generated by Copilot at 4f52115</samp>

This pull request introduces several changes to the server component of Gitpod, mainly related to the authorization and organization management features. It refactors some classes and services, updates the tests and the schema, and fixes some bugs. It also adds a new utility function for resetting the database in the test files.
​

### <samp>🤖 Generated by Copilot at 4f52115</samp>

> _Sing, O Muse, of the valiant pull request_
> _That refactored the code of `IamSessionApp`_
> _And made the tests more robust and clear_
> _With `resetDB`, the helper of the gods._

### <samp>🤖 Generated by Copilot at 4f52115</samp>

*  Add `initialize` method to `Authorizer` class that assigns the installation admin role to the built-in admin user ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-63e2ba5a68d9965170e174c27650c4caed4ac91ad2e79b885e6f9f81655fe5ffR21-R22),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-63e2ba5a68d9965170e174c27650c4caed4ac91ad2e79b885e6f9f81655fe5ffL27-R36))
*  Rename `TeamDB` to `OrganizationService` and update its methods and interface ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-7675ac4e44d727152bb60866081791113aa49e9221fd86001f5df96947f69f18L21-R22),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-7675ac4e44d727152bb60866081791113aa49e9221fd86001f5df96947f69f18L58-R65),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-cc5e8c8b2a81e5a2a660fe5f6a04c00901e4c6152d70e93beb89f49778ef4272L15-R28),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bL7-R8),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bL117-R120),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bL128-R127),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a6a96347dfcaea576e1fc89c848e5a6bcc808f9afb4b071ea4bfe7583c78267aL7-R7),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a6a96347dfcaea576e1fc89c848e5a6bcc808f9afb4b071ea4bfe7583c78267aL174-R175),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a6a96347dfcaea576e1fc89c848e5a6bcc808f9afb4b071ea4bfe7583c78267aL187-R217),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-d6fdb9bb15400f4fbba00db81b2ce6cb7f7bb566dd97b93b9fd4e3fc9a79f479L2848-R2848),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-d6fdb9bb15400f4fbba00db81b2ce6cb7f7bb566dd97b93b9fd4e3fc9a79f479L3500-R3500))
*  Remove `BUILTIN_INSTLLATION_ADMIN_USER_ID` from `User` interface and update imports and references ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-7675ac4e44d727152bb60866081791113aa49e9221fd86001f5df96947f69f18L21-R22),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-cc5e8c8b2a81e5a2a660fe5f6a04c00901e4c6152d70e93beb89f49778ef4272L15-R28),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bL7-R8),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a6a96347dfcaea576e1fc89c848e5a6bcc808f9afb4b071ea4bfe7583c78267aL7-R7))
*  Make internal methods of `IamSessionApp` class private and simplify logic of `createNewOIDCUser` method ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-cc5e8c8b2a81e5a2a660fe5f6a04c00901e4c6152d70e93beb89f49778ef4272L56-R59),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-cc5e8c8b2a81e5a2a660fe5f6a04c00901e4c6152d70e93beb89f49778ef4272L79-R82),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-cc5e8c8b2a81e5a2a660fe5f6a04c00901e4c6152d70e93beb89f49778ef4272L91-R94),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-cc5e8c8b2a81e5a2a660fe5f6a04c00901e4c6152d70e93beb89f49778ef4272L115-R118),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-cc5e8c8b2a81e5a2a660fe5f6a04c00901e4c6152d70e93beb89f49778ef4272L129-R132),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-cc5e8c8b2a81e5a2a660fe5f6a04c00901e4c6152d70e93beb89f49778ef4272L142-R145))
*  Update logic of `addOrUpdateMember` method of `OrganizationService` class to handle first non-admin member and remove admin user from organization ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a6a96347dfcaea576e1fc89c848e5a6bcc808f9afb4b071ea4bfe7583c78267aL174-R175),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a6a96347dfcaea576e1fc89c848e5a6bcc808f9afb4b071ea4bfe7583c78267aL187-R217))
*  Update schema file for `SpiceDBAuthorizer` class to rename `user` relation to `member` and update permissions and assertions ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-69999e6d583968d5b2b70775d0dc480a0a0e4fe001ed7c8eb725518bf320303dL12-R16),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-69999e6d583968d5b2b70775d0dc480a0a0e4fe001ed7c8eb725518bf320303dL36-R36),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-69999e6d583968d5b2b70775d0dc480a0a0e4fe001ed7c8eb725518bf320303dL67-R74),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-69999e6d583968d5b2b70775d0dc480a0a0e4fe001ed7c8eb725518bf320303dL103-R104),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-69999e6d583968d5b2b70775d0dc480a0a0e4fe001ed7c8eb725518bf320303dR150-R151),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-69999e6d583968d5b2b70775d0dc480a0a0e4fe001ed7c8eb725518bf320303dL169-L170))
*  Update `previewctl` command to update existing secret instead of creating a new one in `createAdminCredentials` function ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-e54d16b8ebe207eec554008726e68661aadf73c8fcc0d2e6e99c5b4c41860510L133-R133),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-e54d16b8ebe207eec554008726e68661aadf73c8fcc0d2e6e99c5b4c41860510L142-R143))
*  Add `resetDB` function to clear database tables in test files and update imports and references ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bL14-R17),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-05780e0f6ab3166c26a42575490fa18c16eee36a67cb4d62264be1e4625d2995L7-R7),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-05780e0f6ab3166c26a42575490fa18c16eee36a67cb4d62264be1e4625d2995R28),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-05780e0f6ab3166c26a42575490fa18c16eee36a67cb4d62264be1e4625d2995L74-R74),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-77482d7b22000fadc710e0af7e6b89c599285aab8e7fe4a5f6db0fb3112398c2L7-R7),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-77482d7b22000fadc710e0af7e6b89c599285aab8e7fe4a5f6db0fb3112398c2R18),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-77482d7b22000fadc710e0af7e6b89c599285aab8e7fe4a5f6db0fb3112398c2L56-R56),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-3cc10f050fa6ef9c8caf51f435704d8920868a2f7d551d10ac0832953e917b5cR1-R20),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-92cdccf6d6ba3466df6c81706f564c48dcc249820ded2225071db35f5b1bd323L12-R13),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-92cdccf6d6ba3466df6c81706f564c48dcc249820ded2225071db35f5b1bd323L30-R31))
*  Update test cases to reflect changes in `OrganizationService` methods and logic ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-7675ac4e44d727152bb60866081791113aa49e9221fd86001f5df96947f69f18L58-R65),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-7675ac4e44d727152bb60866081791113aa49e9221fd86001f5df96947f69f18L93-R88),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-7675ac4e44d727152bb60866081791113aa49e9221fd86001f5df96947f69f18L117-R112),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bL48-R56),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bL117-R120),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bL128-R127),[link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-a7b0cc6b128e725c8acf08bd1a3695893b52f25d702a9611734196fa5916bd9bR194-R205))
*  Add check to `expectError` function in `ProjectsService` test file to ensure error has an error code ([link](https://github.com/gitpod-io/gitpod/pull/18280/files?diff=unified&w=0#diff-77482d7b22000fadc710e0af7e6b89c599285aab8e7fe4a5f6db0fb3112398c2R210-R212))

## Related Issue(s)

Fixes WEB-624

## How to test

Setup the preview env with SSO and verify that the first user becomes an owner and has owner privileges with spicedb as well.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - se-use-org-service</li>
	<li><b>🔗 URL</b> - <a href="https://se-use-org-service.preview.gitpod-dev.com/workspaces" target="_blank">se-use-org-service.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - se-use-org-service-gha.13528</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [x] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

**Comment by geropl** [comment]
Taking a look now :eyes:

File: components/server/src/orgs/organization-service.ts [hunk 1]
```diff
@@ -187,12 +188,31 @@ export class OrganizationService {
 
         try {
             await this.teamDB.transaction(async (db) => {
+                const members = await this.teamDB.findMembersByTeam(orgId);
+                const firstMember = members.filter((m) => m.userId !== BUILTIN_INSTLLATION_ADMIN_USER_ID).length === 0;
+                if (firstMember) {
+                    // first member (that is not an admin) is going to be an owner
+                    role = "owner";
+                    log.info({ userId: memberId }, "First member of organization, setting role to owner.");
+                }
```
Comment (geropl) [inline components/server/src/orgs/organization-service.ts:197]: I'm seeing either:
 - the "name your org" screen
 - an error "can't establish a session" (or similar)

Papi logs say: "The user is not currently a member of this organization"


I think we're missing a `await this.teamDB.addMemberToTeam(memberId, orgId);` here.

**Review by geropl — DISMISSED** [review]
https://github.com/gitpod-io/gitpod/pull/18280/files#r1263494172

File: components/server/src/orgs/organization-service.ts [hunk 2]
```diff
@@ -187,12 +188,31 @@ export class OrganizationService {
...
                 await db.setTeamMemberRole(memberId, orgId, role);
                 await this.auth.addOrganizationRole(orgId, memberId, role);
```
Comment (geropl) [inline components/server/src/orgs/organization-service.ts:199]: Unrelated to this PR, but: I just realized it seems we never take away your `owner` role here, even if we switch from `owner` to `member` :grimacing:
@svenefftinge Do we have a test case for that? :thinking:

File: components/server/src/orgs/organization-service.ts (hunk 1 above)
Comment (svenefftinge) [reply components/server/src/orgs/organization-service.ts:197]: fixed

File: components/server/src/orgs/organization-service.ts (hunk 2 above)
Comment (svenefftinge) [reply components/server/src/orgs/organization-service.ts:200]: good catch. fixed and tested

**Comment by svenefftinge** [comment]
copilot:summary

File: components/spicedb/schema/schema.yaml [hunk 3]
```diff
@@ -147,6 +147,8 @@ assertions:
     - organization:org_1#delete@user:user_admin
     - organization:org_1#write_settings@user:user_admin
     - organization:org_1#write_git_provider@user:user_admin
+    # installation admin can create an org
```
Comment (AlexTugarev) [inline components/spicedb/schema/schema.yaml:150]: 👍🏻

**Review by AlexTugarev — APPROVED** [review]
Code looks goods 👍🏻 
Let me test this change on a preview env first, then land it.

**Comment by AlexTugarev** [comment]
Screenshot 2023-07-17 at 17 19 42

Finally, I could verify this works as expected in a preview env.

**Pull request merged**